mediatek: filogic: add Netcore N60 Pro support

libiunc opened the pull request:
Hardware specification:
SoC: MediaTek MT7986A 4x A53
Flash: ESMT F50L1G41LB 128MB
RAM: M16U4G16256A DDR4 512MB
Ethernet: 2x 2.5G + 3x 1G
USB: 1x USB 3.0
WiFi1: MT7975N 2.4GHz 4T4R
WiFi2: MT7975PN 5GHz 4T4R
Button: Reset, WPS
Power: DC 12V 2A

Flash instructions:

   1. Connect to the router using ssh or telnet,
    username: useradmin, password is the web
    login password of the router.
   2. Use scp to upload bl31-uboot.fip and flash:
    "mtd write xxx-bl31-uboot.fip FIP"
    "mtd erase ubi"
   3. Connect to the router via the Lan port,
    set a static ip of your PC.
    (ip 192.168.1.254, gateway 192.168.1.1)
   4. Download initramfs image, reboot router,
    waiting for tftp recovery to complete.
   5. After openwrt boots up, perform sysupgrade.

Note:

1. Back up all mtd partitions before flashing.

Signed-off-by: Yujie Zhu <libriunc@gmail.com>

abajk commented:
Do not open new MRs after each change. This makes it difficult to track whether comments have been addressed.

abajk commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,360 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			led@0 {
```
Looking at the previous MR the LEDs are connected to other pins. Does this device have LEDs on the Ethernet ports at all?

libiunc replied:
> Looking at the previous MR the LEDs are connected to other pins. Does this device have LEDs on the Ethernet ports at all?

This device has only one green LED per network port.It will light up when the 1g or 2.5g speed is connected

abajk replied:
Previously you had configured LED number 3. You can control the state of the LED in LUCI. Are you sure that this diode has the number 0?

libiunc replied:
Modified

libiunc commented:
The strange thing is that I have a signature on my commit, but the github action reports an error saying that is not have signature.

abajk commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -70,6 +70,9 @@ mercusys,mr90x-v1-ubi)
 	ucidef_set_led_netdev "eth0" "LAN" "green:lan" "eth0"
 	;;
+netcore,n60-pro)
+	ucidef_set_led_netdev "wan" "WAN" "blue:wan" "eth1" "link"
```
Here you should add triggers for the LEDs in the PHY.

abajk replied:
Look below at the Zyxel EX5601 as an example.

libiunc replied:
> Here you should add triggers for the LEDs in the PHY.

After removing the LED, the green LED of the network port flashes when there is a network link (netdev). In other words, this phy does not control any LED

abajk replied:
Maybe the two green LEDs of the 2.5G Ethernet ports are controlled by the SoC?

abajk commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,360 @@
+			#size-cells = <0>;
+
+			led@0 {
+				reg = <0>;
```
Does the WAN port have LEDs, too?

libiunc replied:
> Does the WAN port have LEDs, too?

Each network port has only one LED. According to my current changes (pr), it will flash when in link state, regardless of whether it is 2.5g or 1g link speed.There is no 2.5G connection light on this device

libiunc replied:
![Screenshot_2025-03-01-13-09-52-906_com bilibili app blue-edit](https://github.com/user-attachments/assets/66cacf27-9641-4338-8080-4cc3851bfe3f)
![Screenshot_2025-03-01-13-10-06-265_com bilibili app blue-edit](https://github.com/user-attachments/assets/45b6ce46-488e-4ad0-a003-69dd6d0ce2c1)
![Screenshot_2025-03-01-13-10-17-631_com radolyn ayugram-edit](https://github.com/user-attachments/assets/da3dedd8-9c5d-4d42-9e35-905c82551c30)

libiunc replied:
I have corrected it according to your method.

RolandoMagico commented:
> The strange thing is that I have a signature on my commit, but the github action reports an error saying that is not have signature.

Looks like your name in the commit itself doesn't match the name in the signature of the commits.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,338 @@
+	leds {
+		compatible = "gpio-leds";
+
+		sys_status: status {
```
`led_power: power`

aiamadeus commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -1203,6 +1203,31 @@ define Device/netcore_n60
+  IMAGE/sysupgrade.itb := append-kernel | \
+        fit gzip $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb external-static-with-rootfs | append-metadata
+  DEVICE_PACKAGES := kmod-usb3 kmod-mt7915e kmod-mt7986-firmware mt7986-wo-firmware \
+	automount f2fsck mkf2fs
```
Why need `automount f2fsck mkf2fs`?

DragonBluep commented on package/boot/uboot-mediatek/patches/461-add-netcore-n60-pro.patch:
```diff
@@ -0,0 +1,387 @@
+--- /dev/null
++++ b/configs/mt7986_netcore_n60-pro_defconfig
+@@ -0,0 +1,127 @@
++CONFIG_ARM=y
```
Please add symbol `CONFIG_SERIAL_RX_BUFFER=y`. ref: https://github.com/openwrt/openwrt/commit/3d8d807373746d6b3c1674bc450d78090954346f

abajk commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -70,6 +70,12 @@ mercusys,mr90x-v1-ubi)
 	ucidef_set_led_netdev "eth0" "LAN" "green:lan" "eth0"
 	;;
+netcore,n60-pro)
+	ucidef_set_led_netdev "wanlink" "WANLINK" "mdio-bus:06:blue:wan" "eth1" "link tx rx"
```
The LEDs from the PHY are green.

abajk commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,359 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			led@3 {
```
Kernel docs recommends the following pattern "(^led-[0-9a-f]$|led)". Can we use it?

```suggestion
			led-3 {
```

libriunk commented:
I think there is something wrong with the naming of LEDs

hauke commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,359 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&wf_2g_5g_pins>;
+
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
Please use:
```
                       nvmem-cells = <&eeprom_factory_0>;
                       nvmem-cell-names = "eeprom";
```
instead. It also needs some more modifications, see eb510d21588cfa48cb70e696469be03f2d7404e1 for example.

hauke replied:
After doing this change please confirm that the wifi interface still has the correct MAC address.

libiunc replied:
I will deal with this later. I have some problems with my computer.

libiunc replied:
Done

libiunc replied:
![Screenshot_2025-03-20-02-51-12-070_com radolyn ayugram-edit](https://github.com/user-attachments/assets/622c5038-8773-4688-ab8e-3c7ce7532510)

aiamadeus commented on package/boot/uboot-mediatek/patches/461-add-netcore-n60-pro.patch:
```diff
@@ -0,0 +1,388 @@
++		power_led: led-0 {
++			label = "blue:power";
++			gpios = <&gpio 29 GPIO_ACTIVE_LOW>;
++		};
```
Leave a blank line here.
No need for power_led aliases.

aiamadeus commented on package/boot/uboot-mediatek/patches/461-add-netcore-n60-pro.patch:
```diff
@@ -0,0 +1,388 @@
++bootfile_fip=openwrt-mediatek-filogic-netcore_n60-pro-bl31-uboot.fip
++bootfile_upg=openwrt-mediatek-filogic-netcore_n60-pro-squashfs-sysupgrade.itb
++bootled_pwr=blue:power
++bootled_rec=blue:power
```
Can this be merged into one bootled?

libiunc replied:
I think it is better to separate the configuration

aiamadeus replied:
But it’s actually the same LED, no need to differentiate.

libiunc replied:
> But it’s actually the same LED, no need to differentiate.

Which one should I keep?
rec?

aiamadeus replied:
Only use `bootled`, s/bootled_pwr/bootled, s/bootled_rec/bootled

aiamadeus commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -70,6 +70,12 @@ mercusys,mr90x-v1-ubi)
 	;;
+netcore,n60-pro)
+	ucidef_set_led_netdev "wanlink" "WANLINK" "mdio-bus:06:green:wan" "eth1" "link_10 link_100 link_1000 link_2500 tx rx"
+	ucidef_set_led_netdev "lan-1" "lan-1" "mdio-bus:05:green:lan" "lan1" "link_10 link_100 link_1000 link_2500 tx rx"
```
Can this be simplified to "link tx rx"?

libiunc replied:
As far as I know, the LED registered by the current network card driver does not support link only link_<speed>

aiamadeus replied:
But it’s actually the same LED, no need to differentiate.

libiunc replied:
> Can this be simplified to "link tx rx"?

Follow your suggestion. (link tx rx) to change the LED rule on the LAN-1 network port. It will only light up when connected at a 2.5g rate. Other trigger conditions will not make the LED respond.

libiunc replied:
I'll test it again

aiamadeus commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,368 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			led-3 {
```
https://github.com/openwrt/openwrt/commit/a51fd20e0d65ef836ee6c1e13c39a529855af4a0

libiunc replied:
done

libiunc commented:
Reference [from](https://github.com/1715173329/openwrt-official/tree/master-mediatek-n60pro) and make some modifications ([mdio-leds](https://github.com/openwrt/openwrt/commit/a51fd20e0d65ef836ee6c1e13c39a529855af4a0))

aiamadeus commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -70,6 +70,12 @@ mercusys,mr90x-v1-ubi)
 	;;
+netcore,n60-pro)
+	ucidef_set_led_netdev "wanlink" "WANLINK" "mdio-bus:06:green:wan" "eth1" "link_10 link_100 link_1000 link_2500 tx rx"
+	ucidef_set_led_netdev "lan-1" "lan-1" "mdio-bus:05:green:lan" "lan1" "link_10 link_100 link_1000 link_2500 tx rx"
```
https://github.com/openwrt/openwrt/commit/63d56af6c6567f08e18e1f1c8a69f1fac263235b I confirmed this works on n60, maybe you got something wrong.

aiamadeus commented on target/linux/mediatek/filogic/base-files/etc/board.d/01_leds:
```diff
@@ -79,6 +79,11 @@ mercusys,mr90x-v1-ubi)
 	;;
+netcore,n60-pro)
+	ucidef_set_led_netdev "wanlink" "WANLINK" "mdio-bus:06:green:wan" "eth1" "link_10 link_100 link_1000 link_2500 tx rx"
+	ucidef_set_led_netdev "lan-1" "LAN-1" "mdio-bus:05:green:lan" "lan1" "link_10 link_100 link_1000 link_2500 tx rx"
```
Could you please confirm this again?https://github.com/openwrt/openwrt/commit/63d56af6c6567f08e18e1f1c8a69f1fac263235b this works on n60.

abajk replied:
You can shorten the trigger list to “link tx rx”.

libiunc replied:
> Could you please confirm this again?[63d56af](https://github.com/openwrt/openwrt/commit/63d56af6c6567f08e18e1f1c8a69f1fac263235b) this works on n60.

According to your suggestion

The trigger is the same as link tx rx and front led (WAN)

After flashing the firmware or restoring the settings, only the front blue light responds

The led settings in luci show that wanlink is bound to power:blue

Must manually correct it to light up normally

According to the changes you provided, the expected behavior is obtained
The led on the network port will light up when and only when there is a 2.5g link, otherwise link txrx link_* will not light up. I suspect there is a conflict in the trigger here.

aiamadeus replied:
> The led on the network port will light up when and only when there is a 2.5g link, 
> otherwise link txrx link_* will not light up. I suspect there is a conflict in the trigger here.

Maybe it's not connected to led3.

libiunc replied:
> > The led on the network port will light up when and only when there is a 2.5g link,
> > otherwise link txrx link_* will not light up. I suspect there is a conflict in the trigger here.
> 
> Maybe it's not connected to led3.

But this configuration works fine.

aiamadeus replied:
> But this configuration works fine.

I'm a little confused, if it's working properly then the configuration
 "link tx rx" should also light up at any speed.
The code allows only configuration link (without speed) :
https://github.com/torvalds/linux/blob/master/drivers/net/phy/mxl-gpy.c#L952-L966

libiunc replied:
> > But this configuration works fine.
> 
> I'm a little confused, if it's working properly then the configuration "link tx rx" should also light up at any speed. The code allows only configuration link (without speed) : https://github.com/torvalds/linux/blob/master/drivers/net/phy/mxl-gpy.c#L952-L966

The problem here should be that different LEDs cannot share the same trigger(link rx tx). If applied on 1st boot, the first configured one will not work.

aiamadeus replied:
I can't reproduce your problem on other boards with GPY211C. But I think it can be simplified to this:
```
	ucidef_set_led_netdev "lan1" "LAN1" "mdio-bus:05:green:lan" "lan1" "link tx rx"
	ucidef_set_led_netdev "wan" "WAN" "mdio-bus:06:green:wan" "eth1" "tx rx" / or "link tx rx"
	ucidef_set_led_netdev "internet" "Internet" "blue:internet" "eth1" "link"
```

libiunc replied:
> I can't reproduce your problem on other boards with GPY211C. But I think it can be simplified to this:
> 
> ```
> 	ucidef_set_led_netdev "lan1" "LAN1" "mdio-bus:05:green:lan" "lan1" "link tx rx"
> 	ucidef_set_led_netdev "wan" "WAN" "mdio-bus:06:green:wan" "eth1" "tx rx" / or "link tx rx"
> 	ucidef_set_led_netdev "internet" "Internet" "blue:internet" "eth1" "link"
> ```

done.After testing again, it is consistent with the expected behavior of the 1g port.🤔

libiunc replied:
done

libiunc replied:
Is this interface naming based on special led priority (mdio>gpio, like the openwrt one) or in A to z order?🤔

aiamadeus replied:
> Is this interface naming based on special led priority or in A to z order?🤔

Just alphabetical order.

aiamadeus commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,374 @@
+					};
+
+					macaddr_factory_1fef20: macaddr@1fef20 {
+						compatible = "mac-base";
```
No need for "mac-base", so no need for "0>;"

aiamadeus commented on target/linux/mediatek/dts/mt7986a-netcore-n60-pro.dts:
```diff
@@ -0,0 +1,374 @@
+					};
+
+					macaddr_factory_1fef26: macaddr@1fef26 {
+						compatible = "mac-base";
```
Same here.

libiunc committed:
mediatek: filogic: add Netcore N60 Pro support

Hardware specification:
SoC: MediaTek MT7986A 4x A53
Flash: ESMT F50L1G41LB 128MB
RAM: M16U4G16256A DDR4 512MB
Ethernet: 2x 2.5G + 3x 1G
USB: 1x USB 3.0
WiFi1: MT7975N 2.4GHz 4T4R
WiFi2: MT7975PN 5GHz 4T4R
Button: Reset, WPS
Power: DC 12V 2A

Flash instructions:

    Connect to the router using ssh or telnet,
    username: useradmin, password is the web
    login password of the router.
    Use scp to upload bl31-uboot.fip and flash:
    "mtd write xxx-bl31-uboot.fip FIP"
    "mtd erase ubi"
    Connect to the router via the Lan port,
    set a static ip of your PC.
    (ip 192.168.1.254, gateway 192.168.1.1)
    Download initramfs image, reboot router,
    waiting for tftp recovery to complete.
    After openwrt boots up, perform sysupgrade.

Note:

    Back up all mtd partitions before flashing.

Signed-off-by: Yujie Zhu <libriunc@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18138
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
